File: practicas/practica4/src/main/java/deepspace/NumericDamage.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package deepspace;

import java.util.ArrayList;

/**
 *
 * @author inmagalvez
 */
public class NumericDamage extends Damage{
    
    private int nWeapons;
    
    NumericDamage(int w, int s){
        super(s);
        nWeapons = w;
    }
    
    NumericDamage(NumericDamage other){
        this(other.nWeapons, other.getNShields());
    }
    
    @Override
    Damage copy(){
        return new NumericDamage(this);
    }
    
    @Override
    public Damage adjust(ArrayList<Weapon> w, ArrayList<ShieldBooster> s) {
        int shields = adjustShields(s);
        return new NumericDamage(Integer.min(nWeapons, w.size()), shields);
    }
    
    @Override
    public void discardWeapon(Weapon w) {
        if(nWeapons > 0){
            nWeapons--;
        }
    }
    
    @Override
    public boolean hasNoEffect() {
        return (nWeapons == 0 && super.hasNoEffect());
    }
    
    public int getNWeapons() {
        return nWeapons;
    }
    
    
    
    @Override
    public String toString() {
        return super.toString() + "\nnWeapons: " + getNWeapons();
    }
    
    @Override
    DamageToUI getUIversion() {
        return new NumericDamageToUI(this);
    }
     
    
    
}

File: practicas/practica4_rb/lib/NumericDamage.rb
#encoding: utf-8

require_relative 'Damage'
require_relative 'NumericDamageToUI'

module Deepspace
  class NumericDamage < Damage
    public_class_method :new

    attr_reader :nShields, :nWeapons


    def initialize(w, s)
      @nWeapons = w
      super s
    end

    def self.newCopy(d)
      new(d.nWeapons, d.nShields)
    end

    def copy
      NumericDamage.newCopy self
    end

    def discardWeapon(w)
      if nWeapons > 0
        @nWeapons -= 1
      else
        @nWeapons = 0
      end
    end

    def hasNoEffect
      nWeapons.zero? and super
    end

    def adjust(w, s)
      limit_nshields = adjustShields s
      limit_nweapons = [w.length, nWeapons].min
      NumericDamage.new(limit_nweapons, limit_nshields)
    end

    def getUIversion
      NumericDamageToUI.new(self)
    end

    def to_s
      "nWeapons: " + @nWeapons.to_s + "\n" + super
    end

  end
end
